compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs

```diff
@@ -464,6 +464,22 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                                 root_obligation,
                             )
                         } else {
+                            let mut trait_predicate = trait_predicate;
+                            while matches!(
```

Comment:
This doesn't take into account obligation stacks that are like `A -requires> B -requires> C` where `A` and `B` are `#[do_not_recommend]` but `C` is not. I believe we need to basically need to pop `#[do_not_recommend]`s until there are none left in the stack *anywhere*, even if it isn't the *last* obligation in the stack.

For example, consider we put `#[do_not_recommend]` on this `IntoIterator` blanket, and then consider the following code:

```
trait OtherTrait {}

struct It<T>(T);
impl<T> Iterator for It<T> where T: OtherTrait {
    type Item = ();
    fn next(&mut self) -> Option<Self::Item> { todo!() }
}

fn main() {
    for x in It(()) {}
}
```

I would not expect the compiler to mention `(): OtherTrait`, but it would in this case, right? IMO it should just mention `It<()>: IntoIterator`.

From another perspective, considering the full tree of solving a goal like `It<()>: IntoIterator`, we simply wouldn't walk into the obligations that come from considering the blanket implementation marked with `#[do_not_recommend]`.

Reply:
Doesn't that loop not already do that?

We reset the `oblication.cause.code()` field in line 475, so we should pop these obligations as long as we encounter ones that are marked with `#[do_not_recommend]`, right?

Reply:
Not if the top obligation is not a ImplDerivedObligation, right?

Reply:
Thanks for the clarification. That sounds like I should first do some reading on what's actually represented by the `Obligation` struct. Is there any resource on this other than the source code?

Reply:
Hopefully it's now closer to what you imagined, although I'm still not sure about the details of `Obligation`.